stdlib/public/core/Sort.swift

```diff
@@ -249,310 +248,456 @@ extension MutableCollection where Self: RandomAccessCollection {
-        swapAt(b, c)
+    // Storage: [x, x, x, x, x, 6, 8, 8, 10, 12, 15]
+    //           ^              ^
+    //        destLow        srcLow
+    //
+    // Buffer:  [4, 4, 7, 8, 9, x, ...]
+    //           ^              ^
+    //        bufferLow     bufferHigh
+    buffer.moveInitialize(from: low, count: lowCount)
+    bufferHigh = bufferLow + lowCount
+    
+    var srcLow = mid
+
+    // Each iteration moves the element that compares lower into `destLow`,
+    // preferring the buffer when equal to maintain stability. Elements are
+    // moved from either `bufferLow` or `srcLow`, with those pointers
+    // incrementing as elements are moved.
+    while bufferLow < bufferHigh && srcLow < high {
+      if try areInIncreasingOrder(srcLow.pointee, bufferLow.pointee) {
+        destLow.moveInitialize(from: srcLow, count: 1)
+        srcLow += 1
+      } else {
+        destLow.moveInitialize(from: bufferLow, count: 1)
+        bufferLow += 1
+      }
+      destLow += 1
+    }
+  } else {
+    // Move the higher group of elements into the buffer, then traverse from
+    // high to low in both the buffer and the lower group of elements.
+    //
+    // After moving elements, the storage and buffer look like this, where
+    // `x` is uninitialized memory:
+    //
+    // Storage: [4, 4, 7, 8, 9, 6, x, x,  x,  x,  x]
+    //                          ^  ^                 ^
+    //                    srcHigh  destLow        destHigh (past the end)
+    //
+    // Buffer:                    [8, 8, 10, 12, 15, x, ...]
+    //                             ^                 ^
+    //                          bufferLow        bufferHigh
+    buffer.moveInitialize(from: mid, count: highCount)
+    bufferHigh = bufferLow + highCount
+    
+    var destHigh = high
+    var srcHigh = mid
+    destLow = mid
+
+    // Each iteration moves the element that compares higher into `destHigh`,
+    // preferring the buffer when equal to maintain stability. Elements are
+    // moved from either `bufferHigh - 1` or `srcHigh - 1`, with those
+    // pointers decrementing as elements are moved.
+    //
+    // Note: At the start of each iteration, each `...High` pointer points one
+    // past the element they're referring to.
+    while bufferHigh > bufferLow && srcHigh > low {
+      destHigh -= 1
+      if try areInIncreasingOrder(
+        (bufferHigh - 1).pointee, (srcHigh - 1).pointee
+      ) {
+        srcHigh -= 1
+        destHigh.moveInitialize(from: srcHigh, count: 1)
+        
+        // Moved an element from the lower initialized portion to the upper,
+        // sorted, initialized portion, so `destLow` moves down one.
+        destLow -= 1
+      } else {
+        bufferHigh -= 1
+        destHigh.moveInitialize(from: bufferHigh, count: 1)
       }
+    }
+  }
 
-    case (false, true):
-      // 1 swap: 132, 121 --- 2 swaps: 231, 221
-      // swap(b, c): 132->123, 121->112, 231->213, 221->212
-      swapAt(b, c)
+  return true
```

Comment:
Doesn't look like there are any other `return`s here. So perhaps the whole `-> Bool` is not needed?

Reply:
Definitely not needed — it's a workaround for an optimizer issue. Those can all come out once that's resolved.

Reply:
optimizer issues also suggest the benchmarks could improve once resolved...